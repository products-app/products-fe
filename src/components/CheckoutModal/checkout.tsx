import { useState, useEffect, useSyncExternalStore } from 'react'
import { useCartStore } from '@/root/src/store/cart'
import { loadStripe } from '@stripe/stripe-js'
import { Button, Heading, Text } from '@lebernardo/react'
import {
  PaymentElement,
  Elements,
  useStripe,
  useElements,
} from '@stripe/react-stripe-js'
import { getTotal, cartToOrderItems } from '@/helpers/cart'
import { formatDecimalToReal } from '@/helpers/products'
import { toast } from 'react-toastify'
import { postOrder } from '@/api/order'
import { useUserStore } from '@/root/src/store/user'
import config from '@/root/config'

const CheckoutForm = () => {
  const cart = useSyncExternalStore(
    useCartStore.subscribe,
    useCartStore.getState,
  )

  const userID = useUserStore((state) => state.userID)
  const stripe = useStripe()
  const elements = useElements()

  const [errorMessage, setErrorMessage] = useState<string | undefined>(
    undefined,
  )
  const [totalCart, setTotalCart] = useState(0)

  useEffect(() => {
    const [errMsg, totalCart] = getTotal(cart.cartItems)

    if (errMsg) {
      setErrorMessage(errMsg)
      toast.error(errMsg, {
        theme: 'dark',
      })
      return
    }
    setTotalCart(totalCart)
  }, [cart])

  const handleCreateOrder = async () => {
    const orderProducts = cartToOrderItems(cart.cartItems)
    if (!orderProducts) {
      setErrorMessage('the cart is empty')
      toast.error('the cart is empty', {
        theme: 'dark',
      })
      return
    }
    const order = {
      items: orderProducts,
      user_id: userID,
      total: totalCart,
    }

    return postOrder(order)
  }

  const handleSubmit = async (event) => {
    event.preventDefault()

    if (elements == null) {
      return
    }

    const { error: submitError } = await elements.submit()
    if (submitError) {
      setErrorMessage(submitError.message)
      return
    }

    let orderReturned
    await handleCreateOrder()
      .then((res) => {
        orderReturned = res?.data
      })
      .catch((e) => {
        toast.error('não foi possível efetuar a compra', {
          theme: 'dark',
        })
      })

    if (!orderReturned) return
    await stripe?.confirmPayment({
      elements,
      clientSecret: orderReturned.client_secret,
      confirmParams: {
        return_url: '/user/orders',
      },
    })

    cart.truncateItems()
  }

  return (
    <form onSubmit={handleSubmit}>
      <Heading className="text-gray900 mb-3" size="md">
        Finalizar a compra
      </Heading>
      <Text
        className="text-gray600 pb-4 mb-4 border-b-2 border-solid border-gray400"
        size="lg"
      >
        Total do pedido: <strong>{formatDecimalToReal(totalCart)}</strong>
      </Text>
      <PaymentElement />
      <Button
        className="bg-base500 p-2 mt-8 w-full"
        type="submit"
        disabled={!stripe || !elements}
      >
        Finalizar pedido
      </Button>
      {errorMessage && <div>{errorMessage}</div>}
    </form>
  )
}

const stripePromise = loadStripe(config.stripeSK)

const Checkout = () => {
  const cart = useSyncExternalStore(
    useCartStore.subscribe,
    useCartStore.getState,
  )
  const [totalCart, setTotalCart] = useState(200)
  useEffect(() => {
    const [errMsg, totalCart] = getTotal(cart.cartItems)

    if (errMsg) {
      toast.error(errMsg, {
        theme: 'dark',
      })
      return
    }
    setTotalCart(totalCart)
  }, [cart])

  const options = {
    mode: 'payment',
    amount: totalCart * 100,
    currency: 'brl',
    appearance: {},
  }

  return (
    <Elements stripe={stripePromise} options={options}>
      <CheckoutForm />
    </Elements>
  )
}

export default Checkout
