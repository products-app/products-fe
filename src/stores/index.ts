import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'
import { createUserSlice } from './slices/userSlice'
import { createCartSlice } from './slices/cartSlice'
import { createAppSlice } from './slices/appSlice'

export const useStore = create(
  persist<app.ICart & app.IUser & app.IAppControl>(
    (set, get) => ({
      ...createUserSlice(set),
      ...createCartSlice(set, get),
      ...createAppSlice(set),
    }),
    {
      name: 'store',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
