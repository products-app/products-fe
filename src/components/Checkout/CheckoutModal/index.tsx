'use client'
import { useStore } from '@/stores'
import Modal from '@/components/common/Modal'
import Checkout from '../CheckoutContainer'

const CheckoutModal = () => {
  const { openCheckout, setOpenCheckout } = useStore((state) => ({
    openCheckout: state.openCheckout,
    setOpenCheckout: state.setOpenCheckout,
  }))

  function toggleModalCheckout(open: boolean) {
    setOpenCheckout(open)
  }

  return (
    <Modal open={openCheckout} setOpen={toggleModalCheckout}>
      <Checkout />
    </Modal>
  )
}

export default CheckoutModal
