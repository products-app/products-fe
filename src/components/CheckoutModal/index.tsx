'use client';
import Modal from 'react-modal'
import Checkout from './checkout'
import { X as CloseIcon } from 'phosphor-react'
import { useSidebarStore } from '@/store/sidebar'

const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    backgroundColor: '#e1e1e1',
  },
  overlay: {
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
  },
}

Modal.setAppElement('body')

const CheckoutModal = () => {
  const openCheckout = useSidebarStore((state) => state.openCheckout)
  const setOpenCheckout = useSidebarStore((state) => state.setOpenCheckout)

  function closeModal() {
    setOpenCheckout(false)
  }

  return (
    <div>
      <Modal
        isOpen={openCheckout}
        onRequestClose={closeModal}
        style={customStyles}
        contentLabel="Checkout"
      >
        <div className="flex justify-end p-2 w-96">
          <button onClick={closeModal}>
            <CloseIcon />
          </button>
        </div>

        <Checkout />
      </Modal>
    </div>
  )
}

export default CheckoutModal
