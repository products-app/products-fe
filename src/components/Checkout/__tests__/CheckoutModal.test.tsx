import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import CheckoutModal from '../CheckoutModal'
import { useStore } from '@/stores'

const initialState = useStore.getState()

describe('Checkout Modal component', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<CheckoutModal />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should render correctly with open checkout modal', () => {
    useStore.setState({ ...initialState, openCheckout: true }, true)
    const { baseElement } = render(<CheckoutModal />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should call the toggle modal function', async () => {
    useStore.setState({ ...initialState, openCheckout: true }, true)
    render(<CheckoutModal />)

    const handleClick = jest.fn()
    const buttonCloseElement = screen.getByRole('button', {
      name: 'close-modal',
      hidden: true,
    })
    buttonCloseElement.addEventListener('click', handleClick, false)

    await userEvent.click(buttonCloseElement)

    expect(handleClick).toHaveBeenCalled()
  })
})
