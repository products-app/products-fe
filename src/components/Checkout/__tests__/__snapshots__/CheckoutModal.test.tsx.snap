// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout Modal component should render correctly 1`] = `
<body>
  <div>
    <div />
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;

exports[`Checkout Modal component should render correctly with open checkout modal 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <div />
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(0, 0, 0, 0.7); z-index: 9999;"
    >
      <div
        aria-label="Checkout"
        aria-modal="true"
        class="ReactModal__Content"
        role="dialog"
        style="position: absolute; top: 50%; left: 50%; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; margin-right: -50%; transform: translate(-50%, -50%); background-color: rgb(225, 225, 225);"
        tabindex="-1"
      >
        <button
          aria-label="close-modal"
          class="c-bzchUs"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="256"
              width="256"
            />
            <line
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
              x1="200"
              x2="56"
              y1="56"
              y2="200"
            />
            <line
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
              x1="200"
              x2="56"
              y1="200"
              y2="56"
            />
          </svg>
        </button>
        <form>
          <h2
            class="c-dgMvMW c-hpqwme c-dgMvMW-kupvTz-size-md"
          >
            Finalizar a compra
          </h2>
          <p
            class="c-dgMvMW c-CfbqO c-dgMvMW-eQRsLC-size-lg"
          >
            Total do pedido: 
            <strong>
              R$Â 200,00
            </strong>
          </p>
          <div />
          <button
            class="c-gJqLDS c-bcxnOA c-gJqLDS-bIqJQp-size-medium"
            disabled=""
            type="submit"
          >
            Finalizar pedido
          </button>
        </form>
      </div>
    </div>
  </div>
</body>
`;
